### PR TITLE
Fix Deterministic blending with Dominant/Recessive doesn't have initial value even if there is no Discrete track

### LINKED ISSUE #92127: In AnimationTree and AnimationPlayer, rapid crossfading in Dominant mode can make inconsistent results.
### Tested versions

4.3.dev6

### System information

Any

### Issue description

This is a regression by https://github.com/godotengine/godot/pull/89389.

As a compensation for allowing Discrete and Continuous blending, the Dominant mode does not reset to the default value for Deterministic blending when Discrete track could take precedence. This is expected behavior, but the detected cases are too much.

In other words, the problem is that even when there is no Discrete track, it does not reset to the Default value, thereby abandoning the role of the Deterministic blending.

https://github.com/godotengine/godot/assets/61938263/cb8c30e1-5f86-4841-8ef5-f6b597bde6e2

At least if there is a Reset track and it has a Continuous track, it works as intended, so it could be a documentation issue. Because without a Reset track, it is not possible to set an exact default value, so Deterministic blending cannot be used effectively.

However, from a usability standpoint, there should be consistent behavior when there is no Reset track, rather than a default that breaks the blend when a float value track is added and there is no Reset track.

### Steps to reproduce

Adjust blend amount rapidly

### Minimal reproduction project (MRP)

[dominant-reset-blend.zip](https://github.com/godotengine/godot/files/15370138/dominant-reset-blend.zip)

### PR DESCRIPTION
**Pull request by TokageItLab**
Follow up #89389. Fixes https://github.com/godotengine/godot/issues/92127.

I listed the behaviors, some of which were problematic:

| Deterministic  | RESET Track | CallbackModeDiscrete | Is init value appeared in animation with lacking some tracks |
| -- | -- | -- | -- |
| true | None (but there is track in another animation)  | Dominant | No, init value may be discrete |
| true | None (but there is track in another animation)  | Recessive | No, init value may be discrete |

Indeed, the Reset track can be Discrete, but if there is no Discrete track in the other animations, it should be considered Continuous. This causes problems in Deterministic blending, where the results may be inconsistent as https://github.com/godotengine/godot/issues/92127.

With this PR, the Dominant/Recessive mode will first check for the presence of a Reset track, and if there is none, it will check for the presence of blended Discrete track. So the list will be:

| Deterministic  | RESET Track | CallbackModeDiscrete | Is init value appeared in animation with lacking some tracks |
| -- | -- | -- | -- |
| true | None (but there is track in another animation)  | Dominant | Yes, if "there is no other discrete track blended" and "continuous track is blended or blend amount is zero" / But not so, it is No since discrete value should be take precedence |
| true | None (but there is track in another animation)  | Recessive | Yes, if "there is no other discrete track blended" and "continuous track is blended or blend amount is zero" / But not so, it is No since discrete value should be take precedence |

### CODE REVIEW

**Comment by TokageItLab** [comment]
I was organizing the documentation on Discrete mode, and since the https://github.com/godotengine/godot/pull/89389 was invading the role of the Deterministic option and generating confusion, I have reorganized the behavior.

This causes #89163 to recur, but that is the intended specification. It needs to be controlled by the Deterministic option.

**Comment by TokageItLab** [comment]
Adding the is_init flag used by A and B instead of init_use_continuous solves this.

I remember conversing with reduz a bit about this at the beginning of 4.0, but the problem with https://github.com/godotengine/godot/issues/92127 in the first place is that the reset animation does not play when the blend amount is 0.

The workaround is to either apply the reset animation even if the blend amount is 0, or to use a flag to apply the reset animation only once. This PR will use the former for `Force Continuous` and the latter for `Dominant` and `Recessive`. This will make the Dominant mode reset occur only once, so the problem in #89163 should not recur.

File: doc/classes/AnimationMixer.xml [hunk 1]
```diff
@@ -361,14 +361,14 @@
...
-			[b]Note:[/b] If a value track has non-numeric type key values, it is internally converted to use [constant ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT] with [constant Animation.UPDATE_DISCRETE].
+			An [constant Animation.UPDATE_DISCRETE] track value takes precedence when blending [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track values and [constant Animation.UPDATE_DISCRETE] track values.
 		</constant>
 		<constant name="ANIMATION_CALLBACK_MODE_DISCRETE_RECESSIVE" value="1" enum="AnimationCallbackModeDiscrete">
-			An [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track value takes precedence when blending the [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track values and the [constant Animation.UPDATE_DISCRETE] track values.
+			An [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track value takes precedence when blending the [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track values and the [constant Animation.UPDATE_DISCRETE] track values.  This is the default behavior for [AnimationPlayer].
```
Comment (AThousandShips) [inline doc/classes/AnimationMixer.xml:367]:
```suggestion
			An [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track value takes precedence when blending the [constant Animation.UPDATE_CONTINUOUS] or [constant Animation.UPDATE_CAPTURE] track values and the [constant Animation.UPDATE_DISCRETE] track values. This is the default behavior for [AnimationPlayer].
```

File: scene/animation/animation_mixer.cpp [hunk 2]
```diff
@@ -1742,10 +1744,22 @@ void AnimationMixer::_blend_apply() {
...
 				TrackCacheValue *t = static_cast<TrackCacheValue *>(track);
 
-				if (!t->is_variant_interpolatable || !t->use_continuous || (callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT && t->use_discrete)) {
+				if (t->use_discrete && !t->use_continuous) {
+					t->is_init = true; // If only disctere value is applied, no more RESET.
+				}
+
+				if ((t->is_init && (is_zero_amount || !t->use_continuous)) ||
+						(callback_mode_discrete != ANIMATION_CALLBACK_MODE_DISCRETE_FORCE_CONTINUOUS && !is_zero_amount &&
+								(callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT && t->use_discrete))) {
```
Comment (AThousandShips) [inline scene/animation/animation_mixer.cpp:1753]:
```suggestion
								callback_mode_discrete == ANIMATION_CALLBACK_MODE_DISCRETE_DOMINANT && t->use_discrete)) {
```
Unnecessary

**Comment by TokageItLab** [comment]
Now I make Recessive the default due to reevaluating the overall behavior.

In the past, we changed it once in #88492 because there was a problem with the Reset track taking precedence, but this should not be a problem now, and Recessive should be a better default because it does not have the flicker that can occur with Dominant, and it gives the same results as 3.x.

In there is 2 animations:

https://github.com/godotengine/godot/assets/61938263/a967ece8-9d2c-46c6-9d9d-f4178f400889

https://github.com/godotengine/godot/assets/61938263/71d11caf-0c92-44d8-a9cb-b94f3c5cdee0

Force Continuous:

https://github.com/godotengine/godot/assets/61938263/9bcb86b2-d688-45b6-9755-db5cb5119b12

Dominant:

https://github.com/godotengine/godot/assets/61938263/83d4a9b7-0b74-43c9-867e-3a401ef7576f

Recessive:

https://github.com/godotengine/godot/assets/61938263/846adf88-e43c-41c5-acd7-45568611dcdd

**Review by fire — APPROVED** [review]
We're not able to do a good review. We trust Tokage.

**Comment by TokageItLab** [comment]
I would like to include this in beta1 as this changes the default value.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**